src/nupic/algorithms/SDRClassifier.cpp

```diff
@@ -21,489 +21,211 @@
-
-      // update weights
-      if (binary_search(steps_.begin(), steps_.end(), nSteps)) {
-        const vector<Real64> error = calculateError_(bucketIdxList, learnPatternNZ, nSteps);
-        Matrix& w = weightMatrix_.at(nSteps);
-	NTA_ASSERT(alpha_ > 0.0);
-        for (const auto& bit : learnPatternNZ) {
-          for(size_t i = 0; i < error.size(); i++) {
-            const auto val = get_(w, bit, i) + alpha_ * error[i];
-	    if(val == 0) continue;
-	    w[bit][i] = val;
-          }
-        }
-      }
-    }
-  }
+PDF Classifier::infer(const SDR & pattern)
+{
+  // Check input dimensions, or if this is the first time the Classifier has
+  // been used then initialize it with the given SDR's dimensions.
+  if( dimensions_.empty() ) {
+    dimensions_ = pattern.dimensions;
+    while( weights_.size() < pattern.size ) {
+      weights_.push_back( vector<Real>( numCategories_, 0.0f ));
+    }
+  } else if( pattern.dimensions != dimensions_ ) {
+      stringstream err_msg;
+      err_msg << "Classifier input SDR.dimensions mismatch: previously given SDR with dimensions ( ";
+      for( auto dim : dimensions_ )
+        { err_msg << dim << " "; }
+      err_msg << "), now given SDR with dimensions ( ";
+      for( auto dim : pattern.dimensions )
+        { err_msg << dim << " "; }
+      err_msg << ").";
+      NTA_THROW << err_msg.str();
+  }
+
+  // Accumulate feed forward input.
+  PDF probabilities( numCategories_, 0.0f );
+  for( const auto& bit : pattern.getSparse() ) {
+    for( size_t i = 0; i < numCategories_; i++ ) {
+      probabilities[i] += weights_[bit][i];
+    }
+  }
+
+  // Convert from accumulated votes to probability density function.
+  softmax( probabilities.begin(), probabilities.end() );
+  return probabilities;
 }
 
-void SDRClassifier::infer_(const vector<UInt> &patternNZ,
-                           const vector<Real64> &actValue,
-                           ClassifierResult &result)
+
+void Classifier::learn(const SDR &pattern, const vector<UInt> &categoryIdxList)
 {
-  // Add the actual values to the return value. For buckets that haven't
-  // been seen yet, the actual value doesn't matter since it will have
-  // zero likelihood.
-  vector<Real64> &actValueVector = result[ACTUAL_VALUES];
-  actValueVector.reserve( actualValues_.size() );
-
-  for( size_t i = 0; i < actualValues_.size(); ++i ) {
-    if (actualValuesSet_[i]) {
-      actValueVector.push_back( actualValues_[i] );
-    }
-    else {
-      // if doing 0-step ahead prediction, we shouldn't use any
-      // knowledge of the classification input during inference
-      if( steps_.at(0) == 0 ) {
-        actValueVector.push_back( 0.0f );
-      }
-      else {
-        actValueVector.push_back( actValue[0] );
+  // Check if this is a new category & resize the weights table to hold it.
+  const auto maxCategoryIdx = *max_element(categoryIdxList.begin(), categoryIdxList.end());
+  if( maxCategoryIdx >= numCategories_ ) {
```

Comment:
- same as in infer, can you move this all into initialize? It won't change during lifetime of the classifier. 
- Also, may need some doc on categoryIdxList
  - can we pass the encoder here, to obtain both values?

Reply:
> same as in infer, can you move this all into initialize? It won't change during lifetime of the classifier.

I think it's easier for the user to not need to initialize this.  One less argument to remember.

Reply:
> Also, may need some doc on categoryIdxList

Yes.
In this PR I use the words: "categoryIdx" and bucketIdx" interchangeably (they're the same thing in this context).  I'd like to use only categoryIdx and provide an example of converting an integer into a bucketIdx.

Reply:
did you get to do these?

Reply:
..ok, explained here 
https://github.com/htm-community/nupic.cpp/pull/401/files#diff-d6a5469efcd8e0a1cba95db1b5906037R81